Proposals/NNNN-swift-subprocess.md

```diff
@@ -0,0 +1,882 @@
+1. `Process` expects an explicit `URL` to point to its executable instead of trying to resolve the executable path using the `$PATH` variable. This design adds friction in Swift scripting because developers will have to explicitly look up the path to any executable they wish to run.
+2. `Process` expects developers to work with `Pipe` directly to read process output. It could be confusing to determine the correct `fileHandle` to read.
+3. Instead of `async/await`, `Process` uses blocking methods (`.waitUntilExit()`) and callbacks (`.readabilityHandler`) exclusively. This design leaves developers with the responsibility to manage asynchronicity and can easily introduce a "Pyramid of Doom."
+
+
+## Proposed solution
+
+We propose a new type, `struct Subprocess`, that will eventually replace `Process` as the canonical way to launch a process in `Swift`.
+
+Here's the above script rewritten using `Subprocess`:
+
+```swift
+#!/usr/bin/swift
+
+import FoundationEssentials
+
+let gitResult = try await Subprocess.run(   // <- 0
+    executing: .named("git"),               // <- 1
+    arguments: ["diff", "--name-only"]
+)
+
+var changedFiles = String(
+    data: gitResult.standardOutput!,
+    encoding: .utf8)!
+if changedFiles.isEmpty {
+    changedFiles = "No changed files"
+}
+_ = try await Subprocess.run(
+    executing: .named("say"),
+    arguments: [changedFiles]
+)
+```
+
+Let's break down the example above:
+
+0. `Subprocess` is constructed entirely on the `async/await` paradigm. The `run()` method utilizes `await` to allow the child process to finish, asynchronously returning an `ExecutionResult`. Additionally, there is an closure based overload of `run()` that offers more granulated control, which will be discussed later.
+1. There are two primary ways to configure the executable being run:
+    - The default approach, recommended for most developers, is `.at(FilePath)`. This allows developers to specify a full path to an executable.
+    - Alternatively, developers can use `.named(String)` to instruct `Subprocess` to look up the executable path based on heuristics such as the `$PATH` environment variable.
+
+
+## Detailed Design
+
+### The New `Subprocess` Type
+
+We propose a new `struct Subprocess`. Developers primarily interact with `Subprocess` via the static `run` methods which asynchronously executes a subprocess.
+
+```swift
+@available(FoundationPreview 0.4, *)
+@available(iOS, unavailable)
+@available(tvOS, unavailable)
+@available(watchOS, unavailable)
+extension Subprocess {
+    public static func run(
```

Comment:
@iCharlesHu This needs to specify the cancellation behaviour. For Swift Concurrency to work it's crucial that under task cancellation (`taskGroup.cancelAll()`) the `async` methods return quickly. For sub processes that will mean that it needs to get rid of the process which requires sending a signal -- likely `SIGKILL`. But this needs to be clearly specified.

Reply:
I mentioned this in the [pitch](https://forums.swift.org/t/pitch-swift-subprocess/69805/38) as well and Charles said that he is going to change this to send a SIGKILL